base/compiler/optimize.jl

```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
     return nothing
 end
 
+function visit_bb_phis!(callback, ir::IRCode, bb::Int)
+    stmts = ir.cfg.blocks[bb].stmts
+    for idx in stmts
+        stmt = ir[SSAValue(idx)][:inst]
+        if !isa(stmt, PhiNode)
+            if !is_valid_phiblock_stmt(stmt)
+                return
+            end
+        else
+            callback(idx, stmt)
+        end
+    end
+end
+
+function any_stmt_may_throw(ir::IRCode, bb)
+    for stmt in ir.cfg.blocks[bb].stmts
+        if (ir[SSAValue(stmt)][:flag] & IR_FLAG_NOTHROW) != 0
+            return true
+        end
+    end
+    return false
+end
+
+function conditional_successors_may_throw(lazypostdomtree::LazyPostDomtree, ir::IRCode, bb::Int)
+    visited = BitSet((bb,))
+    worklist = Int[bb]
+    postdomtree = get!(lazypostdomtree)
+    while !isempty(worklist)
+        thisbb = pop!(worklist)
+        for succ in ir.cfg.blocks[thisbb].succs
+            succ in visited && continue
+            push!(visited, succ)
+            postdominates(postdomtree, succ, thisbb) && continue
+            any_stmt_may_throw(ir, succ) && return true
+            push!(worklist, succ)
+        end
+    end
+    return false
+end
+
+function ipo_dataflow_analysis!(interp::AbstractInterpreter, ir::IRCode, result::InferenceResult)
+    inconsistent = BitSet()
+    inconsistent_bbs = BitSet()
+    tpdum = TwoPhaseDefUseMap(length(ir.stmts))
+    lazypostdomtree = LazyPostDomtree(ir)
+
+    all_effect_free = true # TODO refine using EscapeAnalysis
+    all_nothrow = true
+    all_retpaths_consistent = true
+    all_noub = true
+    any_conditional_ub = false
+    had_trycatch = false
+
+    scanner = BBScanner(ir)
+
+    effects = result.ipo_effects
+
+    cfg = nothing
+    domtree = nothing
+    function get_augmented_domtree()
```

Comment:
Probably better to set up a new struct 
```julia
struct AugumentedDomtree
    cfg::CFG
    domtree::DomTree
end
```
and make `get_augumented_domtree` return the struct instead for better inferrability.